pkg/controller/podgroup.go

```diff
@@ -58,19 +119,224 @@ func newPodGroup(mpiJob *kubeflow.MPIJob) *volcanov1beta1.PodGroup {
+}
+
+func (s *SchedulerPluginsCtrl) decoratePodTemplateSpec(pts *corev1.PodTemplateSpec, mpiJobName string) {
+	if pts.Spec.SchedulerName != s.schedulerName {
+		klog.Warningf("%s scheduler is specified when gang-scheduling is enabled and it will be overwritten", pts.Spec.SchedulerName)
 	}
+	pts.Spec.SchedulerName = s.schedulerName
+	if pts.Labels == nil {
+		pts.Labels = make(map[string]string)
+	}
+	pts.Labels[schedv1alpha1.PodGroupLabel] = mpiJobName
 }
 
-// calcPriorityClassName calculates the priorityClass name needed for podGroup according to the following priorities:
+// calculatePGMinResources will calculate minResources for podGroup.
+// It calculates the sum of resources defined in all containers and will return the result.
+// If the number of replicas (.spec.mpiReplicaSpecs[Launcher].replicas + .spec.mpiReplicaSpecs[Worker].replicas)
+// is more of minMember, it reorders replicas according to each priorityClass setting in `podSpec.priorityClassName`
+// and then resources with a priority less than minMember will not be added to minResources.
+// Note that it doesn't account for the priorityClass specified in podSpec.priorityClassName
+// if the priorityClass doesn't exist in the cluster when it reorders replicas.
+//
+// By adding appropriate required resources to a podGroup resource,
+// the coscheduling plugin can filter out the pods that belong to the podGroup in PreFilter
+// if the cluster doesn't have enough resources.
+// ref: https://github.com/kubernetes-sigs/scheduler-plugins/blob/93d7c92851c4a17f110907f3b5be873176628441/pkg/coscheduling/core/core.go#L159-L182
+func (s *SchedulerPluginsCtrl) calculatePGMinResources(minMember *int32, mpiJob *kubeflow.MPIJob) *corev1.ResourceList {
+	if schedPolicy := mpiJob.Spec.RunPolicy.SchedulingPolicy; schedPolicy != nil && schedPolicy.MinResources != nil {
+		return schedPolicy.MinResources
+	}
+	if minMember != nil && *minMember == 0 {
+		return nil
+	}
+
+	var order replicasOrder
+	for rt, replica := range mpiJob.Spec.MPIReplicaSpecs {
+		rp := replicaPriority{
+			priority:    0,
+			ReplicaSpec: *replica,
+		}
+		pcName := replica.Template.Spec.PriorityClassName
+		if len(pcName) != 0 {
+			if priorityClass, err := s.PriorityClassLister.Get(pcName); err != nil {
+				klog.Warningf("Ignore replica %q priority class %q: %v", rt, pcName, err)
+			} else {
+				rp.priority = priorityClass.Value
+			}
+		}
+		order = append(order, rp)
+	}
+
+	sort.Sort(sort.Reverse(order))
+	// Launcher + Worker > minMember
+	if minMember != nil && *order[0].Replicas+*order[1].Replicas > *minMember {
+		// NUM(Workers) = minMember - NUM(Launcher)
+		order[1].Replicas = pointer.Int32(*minMember - 1)
+	}
+
+	minResources := corev1.ResourceList{}
+	for _, rp := range order {
+		if rp.Replicas == nil {
+			continue
+		}
+		for i := int32(0); i < *rp.Replicas; i++ {
+			for _, c := range rp.Template.Spec.Containers {
```

Comment:
what about initContainers?

Reply:
resource calculation in scheduler is: `max(initContainers, sum(containers)) + pod overhead`.

Also, this is ignoring any limit ranges or defaulting.

But I guess it's ok if this is just an estimate. I would guess that the scheduler ultimately considers the final pods.

Reply:
> But I guess it's ok if this is just an estimate. I would guess that the scheduler ultimately considers the final pods.

I think so too. It's fine with only containers since the conscheduling plugin isn't quota management software :)